### PR TITLE
Feat/custom model support

### PR DESCRIPTION
**Pull request by cbrzn**
this PR aims to support local models using the [custom client models](https://microsoft.github.io/autogen/blog/2024/01/26/Custom-Models/) so we can have a succesful integration with morpheus agents. it uses llama cpp (which offers an interface for open source models) to interact with the [model being used](https://huggingface.co/meetkai/functionary-small-v2.4-GGUF) for morpheus agents

### CODE REVIEW

**Review by dOrgJelli — APPROVED** [review]
LGTM

File: autotx/helper_agents/user_proxy.py [hunk 1]
```diff
@@ -1,13 +1,16 @@
...
+if TYPE_CHECKING:
+    from autotx.AutoTx import CustomModel
+
+def build(user_prompt: str, agents_information: str, get_llm_config: Callable[[], Optional[Dict[str, Any]]], custom_model: Optional['CustomModel']) -> UserProxyAgent:
     user_proxy = UserProxyAgent(
         name="user_proxy",
         is_termination_msg=lambda x: x.get("content", "") and "TERMINATE" in x.get("content", ""),
         human_input_mode="NEVER",
-        max_consecutive_auto_reply=10,
+        max_consecutive_auto_reply=4 if custom_model else 10,
```
Comment (nerfZael) [inline autotx/helper_agents/user_proxy.py:13]: Interesting, why this change?

File: autotx/utils/LlamaClient.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    def cost(self, _: Union[ChatCompletion, Completion]) -> float:
+        return 0.0
+
+    def get_usage(self, _: Union[ChatCompletion, Completion]) -> dict[str, Any]:
+        return {
+            "prompt_tokens": 0,
+            "completion_tokens": 0,
+            "total_tokens": 0,
+            "cost": 0,
+            "model": "meetkai/functionary-small-v2.4-GGUF",
```
Comment (nerfZael) [inline autotx/utils/LlamaClient.py:46]: Do we wanna have this hard coded?

File: autotx/utils/LlamaClient.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+        }
+
+    def _sanitize_chat_completion_messages(
```
Comment (nerfZael) [inline autotx/utils/LlamaClient.py:49]: What does this do?

File: autotx/utils/LlamaClient.py (hunk 2 above)
Comment (cbrzn) [reply autotx/utils/LlamaClient.py:46]: oo - nice catch, no we dont

File: autotx/utils/LlamaClient.py (hunk 3 above)
Comment (cbrzn) [reply autotx/utils/LlamaClient.py:49]: when doing a tool call, the message from the chat completion doesn't return the name of the function w/the response - this is what the llamacpp bindings expect
```json
{
  tool_call_id": tool_call.id,
  "role": "tool",
  "name": function_name,
  "content": function_response,
}
```
so we're manually adding the `name` to that object to make things work as expected

File: autotx/helper_agents/user_proxy.py (hunk 1 above)
Comment (cbrzn) [reply autotx/helper_agents/user_proxy.py:13]: the user proxy is not able to consistently terminate when the goal has been done - in order to avoid a lot of time of execution, we put a lower number for auto replies
